samples/nfc/tag_reader/src/main.c

```diff
@@ -467,76 +444,195 @@ static void t4t_isodep_data_send(u8_t *data, size_t data_len, u32_t ftd)
+
+	printk("NFC T4T Capability Container file read.\n");
+
+	for (size_t i = 0; i < cc->tlv_count; i++) {
+		tlv_block = &cc->tlv_block_array[i];
+
+		if ((tlv_block->type == NFC_T4T_TLV_BLOCK_TYPE_NDEF_FILE_CONTROL_TLV) &&
+		    (tlv_block->value.read_access == NFC_T4T_TLV_BLOCK_CONTROL_FILE_READ_ACCESS_GRANTED)) {
+			err = nfc_t4t_hl_procedure_ndef_file_select(tlv_block->value.file_id);
+			if (err) {
+				printk("NFC T4T NDEF file select error: %d.\n", err);
+			}
+
+			return;
+		}
+	}
+
+	printk("No NDEF File TLV in Capability Container.");
+}
+
+static void t4t_hl_ndef_read(u16_t file_id, const u8_t *data, size_t len)
+{
+	int err;
+	struct nfc_t4t_cc_file *cc;
+	struct nfc_t4t_tlv_block *tlv_block;
+
+	printk("NDEF file read, id: 0x%x.\n", file_id);
+
+	t4t.tlv_index++;
+	cc = &NFC_T4T_CC_DESC(t4t_cc);
+
+	for (size_t i = t4t.tlv_index; i < cc->tlv_count; i++) {
+		tlv_block = &cc->tlv_block_array[i];
+
+		if ((tlv_block->type == NFC_T4T_TLV_BLOCK_TYPE_NDEF_FILE_CONTROL_TLV) &&
+		    (tlv_block->value.read_access == NFC_T4T_TLV_BLOCK_CONTROL_FILE_READ_ACCESS_GRANTED)) {
+
+			err = nfc_t4t_hl_procedure_ndef_file_select(tlv_block->value.file_id);
+			if (err) {
+				printk("NFC T4T NDEF file select error: %d.\n", err);
+			}
+
+			return;
+		}
+
+		t4t.tlv_index++;
+	}
+
+	nfc_t4t_cc_file_printout(cc);
+
+	tlv_block = cc->tlv_block_array;
+	for (size_t i = 0; i < cc->tlv_count; i++) {
+		if ((tlv_block[i].type == NFC_T4T_TLV_BLOCK_TYPE_NDEF_FILE_CONTROL_TLV) ||
+		    (tlv_block[i].value.file.content != NULL)) {
+			ndef_data_analyze(tlv_block[i].value.file.content + TAG_TYPE_4_NLEN_FIELD_SIZE,
+					tlv_block[i].value.file.len - TAG_TYPE_4_NLEN_FIELD_SIZE);
+		}
+	}
+
+	t4t.tlv_index = 0;
+
+	err = nfc_t4t_isodep_tag_deselect();
+	if (err) {
+		printk("NFC T4T Deselect error: %d.\n", err);
+	}
+}
+
+static const struct nfc_t4t_hl_procedure_cb t4t_hl_procedure_cb = {
+	.selected = t4t_hl_selected,
+	.cc_read = t4t_hl_cc_read,
+	.ndef_read = t4t_hl_ndef_read
+};
+
 void main(void)
 {
 	int err;
 
-	printk("NFC reader sample started\n");
+	printk("NFC reader sample started.\n");
+	nfc_t4t_hl_procedure_cb_register(&t4t_hl_procedure_cb);
 
 	k_delayed_work_init(&transmit_work, transfer_handler);
 
 	err = nfc_t4t_isodep_init(tx_data, sizeof(tx_data),
 				  t4t.data, sizeof(t4t.data),
 				  &t4t_isodep_cb);
 	if (err) {
-		printk("NFC T4T ISO-DEP Protocol initialization failed err: %d\n",
+		printk("NFC T4T ISO-DEP Protocol initialization failed err: %d.\n",
 		       err);
 		return;
 	}
 
 	err = st25r3911b_nfca_init(events, ARRAY_SIZE(events), &cb);
 	if (err) {
-		printk("NFCA initialization failed err: %d\n", err);
+		printk("NFCA initialization failed err: %d.\n", err);
 		return;
 	}
 
 	err = st25r3911b_nfca_field_on();
 	if (err) {
-		printk("Field on error %d", err);
+		printk("Field on error %d.", err);
 		return;
 	}
 
 	while (true) {
 		k_poll(events, ARRAY_SIZE(events), K_FOREVER);
 		err = st25r3911b_nfca_process();
```

Comment:
quick question? are events reset? (`K_POLL_STATE_NOT_READY` set there). I think you do it in the lib but I want to be sure. Sorry - I am checking on the browser at home and jumping around in github code browser does not exist.

Reply:
Yes. Events are reset inside reader/writer driver. They are set to `K_POLL_STATE_NOT_READY ` inside `st25r3911b_nfca_process` function.